File: Practice/AllDomains/Tutorials/30DaysOfCode/Day12Inheritance.cs
using System;
using System.Linq;

class Person{
	protected string firstName;
	protected string lastName;
	protected int id;
	
	public Person(){}
	public Person(string firstName, string lastName, int identification){
			this.firstName = firstName;
			this.lastName = lastName;
			this.id = identification;
	}
	public void printPerson(){
		Console.WriteLine("Name: " + lastName + ", " + firstName + "\nID: " + id); 
	}
}

class Student : Person
{
	/*	author:@shivkrthakur  */
    private int[] testScores;  
   
   public Student(string firstName, string lastName, int id, int [] scores): base(firstName, lastName, id)
   {
        this.testScores = scores;    
   }
    
    public char calculate()
    {
        int average = Average();
        if(average >= 90 && average <=100) return 'O';
        if(average >= 80 && average < 90) return 'E';
        if(average >= 70 && average < 80) return 'A';
        if(average >= 55 && average < 70) return 'P';
        if(average >= 40 && average < 55) return 'D';
        if(average < 40) return 'T';
        return 'F';
    }
    
    public int Average()
    {
        int sum = 0;
        for(int x = 0; x < testScores.Length; x++)
        {
            sum += testScores[x];
        }
        
        return sum/testScores.Length;
    }
}

class Solution {
	static void Main() {
		string[] inputs = Console.ReadLine().Split();
		string firstName = inputs[0];
	  	string lastName = inputs[1];
		int id = Convert.ToInt32(inputs[2]);
		int numScores = Convert.ToInt32(Console.ReadLine());
		inputs = Console.ReadLine().Split();
	  	int[] scores = new int[numScores];
		for(int i = 0; i < numScores; i++){
			scores[i]= Convert.ToInt32(inputs[i]);
		} 
	  	
		Student s = new Student(firstName, lastName, id, scores);
		s.printPerson();
		Console.WriteLine("Grade: " + s.Calculate() + "\n");
	}
}

File: Practice/AllDomains/Tutorials/30DaysOfCode/Day12Inheritance.js
/*	author:@shivkrthakur  */
'use strict';

var _input = '';
var _index = 0;
process.stdin.on('data', (data) => { _input += data; });
process.stdin.on('end', () => {
    _input = _input.split(new RegExp('[ \n]+'));
    main();
});
function read() { return _input[_index++]; }

/**** Ignore above this line. ****/

class Person {
    constructor(firstName, lastName, identification) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.idNumber = identification;
    }

    printPerson() {
        console.log(
            "Name: " + this.lastName + ", " + this.firstName
            + "\nID: " + this.idNumber
        )
    }
}

class Student extends Person {
    /*
    *   Class Constructor
    *
    *   @param firstName - A string denoting the Person's first name.
    *   @param lastName - A string denoting the Person's last name.
    *   @param id - An integer denoting the Person's ID number.
    *   @param scores - An array of integers denoting the Person's test scores.
    */
    // Write your constructor here
    constructor(firstName, lastName, id, scores){
        super(firstName, lastName, id);
        this.scores = scores;
    }
    /*
    *   Method Name: calculate
    *   @return A character denoting the grade.
    */
    // Write your method here
    calculate(){
        var total = 0;
        for(var i = 0; i < this.scores.length; i++){
            total += this.scores[i];
        }
        total = total/this.scores.length;
        var string = '';
        if(total >= 90 && total <= 100) string = 'O';
        else if(total >= 80 && total < 90) string = 'E';
        else if(total >= 70 && total < 80) string = 'A';
        else if(total >= 55 && total < 70) string = 'P';
        else if(total >= 40 && total < 55) string = 'D';
        else if(total < 40) string = 'T';
        return string;
    }
}

function main() {
    let firstName = read()
    let lastName = read()
    let id = +read()
    let numScores = +read()
    let testScores = new Array(numScores)

    for (var i = 0; i < numScores; i++) {
        testScores[i] = +read()
    }

    let s = new Student(firstName, lastName, id, testScores)
    s.printPerson()
    s.calculate()
    console.log('Grade: ' + s.calculate())
}
